tests/platform/test_reboot.py

```diff
@@ -21,42 +21,109 @@
+REBOOT_TYPE_REBOOT_BY_POWEROFF = "power off"
+REBOOT_TYPE_REBOOT_BY_WATCHDOG = "watchdog"
+
+reboot_ctrl_dict = {
+    REBOOT_TYPE_REBOOT_BY_POWEROFF : {
+        REBOOT_TIMEOUT : 300,
+        REBOOT_CAUSE : "Power Loss"
+    },
+    REBOOT_TYPE_COLDREBOOT : {
+        REBOOT_COMMAND : "reboot",
+        REBOOT_TIMEOUT : 300,
+        REBOOT_CAUSE : "reboot"
+    },
+    REBOOT_TYPE_FASTREBOOT : {
+        REBOOT_COMMAND : "fast-reboot",
+        REBOOT_TIMEOUT : 180,
+        REBOOT_CAUSE : "fast-reboot"
+    },
+    REBOOT_TYPE_WARMREBOOT : {
+        REBOOT_COMMAND : "warm-reboot",
+        REBOOT_TIMEOUT : 180,
+        REBOOT_CAUSE : "warm-reboot"
+    },
+    REBOOT_TYPE_REBOOT_BY_WATCHDOG : {
+        REBOOT_COMMAND : "python -c \"import sonic_platform.platform as P; P.Platform().get_chassis().get_watchdog().arm(5); exit()\"",
+        REBOOT_TIMEOUT : 300,
+        REBOOT_CAUSE : "Watchdog"
+    }
+}
+
+def check_reboot_cause(dut, reboot_cause_expected):
+    """
+    @summary: Check the reboot cause on DUT.
+    @param dut: The AnsibleHost object of DUT.
+    @param reboot_cause_expected: The expected reboot cause.
+    """
+    logging.info("Check the reboot cause")
+    output = dut.shell("show reboot-cause")
+    reboot_cause_got = output["stdout"]
+    logging.debug("show reboot-cause returns {}".format(reboot_cause_got))
+    m = re.search(reboot_cause_expected, reboot_cause_got)
+    assert m is not None, "got reboot-cause %s after rebooted by %s" % (reboot_cause_got, reboot_cause_expected)
+
+
+def reboot_and_check(localhost, dut, interfaces, reboot_type="cold", reboot_helper=None, reboot_argu=None):
     """
     Perform the specified type of reboot and check platform status.
+    @param dut: The AnsibleHost object of DUT.
+    @param interfaces: DUT's interfaces defined by minigraph
+    @param reboot_type: The reboot type, pre-defined const that has name convention of REBOOT_TYPE_XXX.
+    @param reboot_helper: The helper function used only by power off reboot
+    @param reboot_argu: The argument used by reboot_helper
     """
     logging.info("Run %s reboot on DUT" % reboot_type)
-    if reboot_type == "cold":
-        reboot_cmd = "reboot"
-        reboot_timeout = 300
-    elif reboot_type == "fast":
-        reboot_cmd = "fast-reboot"
-        reboot_timeout = 180
-    elif reboot_type == "warm":
-        reboot_cmd = "warm-reboot"
-        reboot_timeout = 180
+
+    reboot_ctrl_element = reboot_ctrl_dict.get(reboot_type)
+    if reboot_ctrl_element is None:
+        assert False, "Unknown reboot type %s" % reboot_type
+
+    reboot_timeout = reboot_ctrl_element[REBOOT_TIMEOUT]
+    reboot_cause = reboot_ctrl_element[REBOOT_CAUSE]
+    if reboot_type == REBOOT_TYPE_REBOOT_BY_POWEROFF:
+        assert reboot_helper is not None, "A reboot function must be provided for power off reboot"
+        if reboot_helper is None:
+            assert False, "reboot_helper must be provided for power off reboot"
```

Comment:
The above two lines of code won't be hit if `reboot_helper` is None. Function duplicated with the line before:
```
        assert reboot_helper is not None, "A reboot function must be provided for power off reboot"
```

Reply:
fixed